crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1134 @@
+                interpreter: Interpreter::query(path, cache)?,
+            }
+        }
+        InterpreterRequest::Directory(path) => {
+            if !sources.contains(InterpreterSource::ProvidedPath) {
+                return Err(Error::SourceNotSelected(
+                    request.clone(),
+                    InterpreterSource::ProvidedPath,
+                ));
+            }
+            if !path.try_exists()? {
+                return Ok(NoInterpreter(InterpreterNotFound::FileNotFound(
+                    path.clone(),
+                )));
+            }
+            let executable = virtualenv_python_executable(path);
+            if !executable.try_exists()? {
+                return Ok(NoInterpreter(
+                    InterpreterNotFound::ExecutableNotFoundInDirectory(path.clone(), executable),
+                ));
+            }
+            DiscoveredInterpreter {
+                source: InterpreterSource::ProvidedPath,
+                interpreter: Interpreter::query(executable, cache)?,
+            }
+        }
+        InterpreterRequest::ExecutableName(name) => {
+            if !sources.contains(InterpreterSource::SearchPath) {
+                return Err(Error::SourceNotSelected(
+                    request.clone(),
+                    InterpreterSource::SearchPath,
+                ));
+            }
+            let Some(executable) = which(name).ok() else {
+                return Ok(NoInterpreter(
+                    InterpreterNotFound::ExecutableNotFoundInSearchPath(name.clone()),
+                ));
+            };
+            DiscoveredInterpreter {
+                source: InterpreterSource::SearchPath,
+                interpreter: Interpreter::query(executable, cache)?,
+            }
+        }
+        InterpreterRequest::Implementation(implementation) => {
+            let Some((source, interpreter)) = python_interpreters(None, sources, cache)
+                .find(|result| {
+                    result.is_err()
+                        || result.as_ref().is_ok_and(|(_source, interpreter)| {
+                            interpreter.implementation_name() == implementation.as_str()
+                        })
+                })
+                .transpose()?
+            else {
+                return Ok(NoInterpreter(
+                    InterpreterNotFound::NoMatchingImplementation(sources.clone(), *implementation),
+                ));
+            };
+            DiscoveredInterpreter {
+                source,
+                interpreter,
+            }
+        }
+        InterpreterRequest::ImplementationVersion(implementation, version) => {
+            let Some((source, interpreter)) = python_interpreters(Some(version), sources, cache)
+                .find(|result| {
+                    result.is_err()
+                        || result.as_ref().is_ok_and(|(_source, interpreter)| {
+                            version.matches_interpreter(interpreter)
```

Comment:
Confirming the logic here -- we take the first interpreter that matches the query, right? So e.g. if the user requested `3.9`, and the first `3.9` we saw was `3.9.0`, we'd accept that (even if we might have `3.9.12` later in `PATH`) -- is that correct? (IMO that is the correct behavior but confirming.)

Reply:
Yep!